PopPUNK/utils.py

```diff
@@ -497,3 +494,162 @@ def isolateNameToLabel(names):
+        fail_grp = hdf_fail.create_group('sketches')
+    seq_length = {}
+    seq_ambiguous = {}
+    seq_excluded = {}
+    removed = []
+    retained = []
+    
+    # iterate through sketches
+    for dataset in read_grp:
+        # test thresholds
+        remove = False
+        seq_length[dataset] = hdf_in['sketches'][dataset].attrs['length']
+        seq_ambiguous[dataset] = hdf_in['sketches'][dataset].attrs['missing_bases']
+        # if no filtering to be undertaken, retain all sequences
+        if qc_filter == 'continue':
+            retained.append(dataset)
+
+    # calculate thresholds
+    with open(prefix + '/' + os.path.basename(prefix) + '_qcreport.txt', 'a+') as qc_file:
+    
+        # get mean length if not using user-supplied estimation
+        if estimated_length is None:
+            genome_lengths = np.fromiter(seq_length.values(), dtype = int)
+            mean_genome_length = np.mean(genome_lengths)
+        
+        # calculate length threshold unless user-supplied
+        if lower_length is None:
+            lower_length = mean_genome_length - length_sigma*np.std(genome_lengths)
+            upper_length = mean_genome_length + length_sigma*np.std(genome_lengths)
+
+        # iterate through and filter
+        example_assembly = None
+        failed_sample = False
+        for dataset in seq_length.keys():
+                    
+            # determine if sequence passes filters
+            remove = False
+            if seq_length[dataset] < lower_length:
+                remove = True
+                qc_file.write(dataset + '\tBelow lower length threshold\n')
+            elif seq_length[dataset] > upper_length:
+                qc_file.write(dataset + '\tAbove upper length threshold\n')
+            if upper_n is not None and seq_ambiguous[dataset] > upper_n:
+                remove = True
+                qc_file.write(dataset + '\tAmbiguous sequence too high\n')
+            elif seq_ambiguous[dataset] > prop_n * seq_length[dataset]:
+                remove = True
+                qc_file.write(dataset + '\tAmbiguous sequence too high\n')
+
+            # write to files
+            if remove:
+                sys.stderr.write(dataset + ' failed QC\n')
+                failed_sample = True
+                if retain_failures:
+                    fail_grp.copy(read_grp[dataset], dataset)
+            else:
+                if qc_filter == 'prune':
+                    out_grp.copy(read_grp[dataset], dataset)
+                    retained.append(dataset)
+                # retain an example assembly
+                if example_assembly is None:
+                    example_assembly = dataset
+    
+        # get kmers from original database
+        db_kmers = hdf_in['sketches'][example_assembly].attrs['kmers']
+        
+        # close files
+        hdf_in.close()
+        if qc_filter == 'prune':
+            hdf_out.close()
+        if retain_failures:
+            hdf_fail.close()
```

Comment:
I've learned that it's good to put all of this in a `try` block, because if there's an error and they're not closed it can corrupt them. So everything after when they're opened should be in a try block, and the except block should also close them before exiting

Reply:
Good to know, done in 66e1cc4.